### PR TITLE
Add ImageAlgorithm to rotate an image

### PR DESCRIPTION
**Pull request by upengareri**

### CODE REVIEW

**Review by AshAswin — CHANGES_REQUESTED** [review]
Please make the requested changes

File: catkin_ws/src/image_rotation/include/image_rotation/ImageAlgorithm.h [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#ifndef IMAGE_ALGORITHM_H
```
Comment (AshAswin) [inline catkin_ws/src/image_rotation/include/image_rotation/ImageAlgorithm.h]: files should be named with snake case words separated by underscores

File: catkin_ws/src/image_rotation/include/image_rotation/ImageAlgorithm.h [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+#define IMAGE_ALGORITHM_H
+
+#include <opencv2/opencv.hpp>
+
+
+class ImageAlgorithm
```
Comment (AshAswin) [inline catkin_ws/src/image_rotation/include/image_rotation/ImageAlgorithm.h]: Consider replacing the class with a namespace to group utility functions. Refer: https://softwareengineering.stackexchange.com/questions/134540/are-utility-classes-with-nothing-but-static-members-an-anti-pattern-in-c , https://stackoverflow.com/questions/7345956/advantages-of-classes-with-only-static-methods-in-c

File: catkin_ws/src/image_rotation/include/image_rotation/ImageAlgorithm.h [hunk 3]
```diff
@@ -0,0 +1,17 @@
+#ifndef IMAGE_ALGORITHM_H
+#define IMAGE_ALGORITHM_H
+
+#include <opencv2/opencv.hpp>
+
```
Comment (AshAswin) [inline catkin_ws/src/image_rotation/include/image_rotation/ImageAlgorithm.h]: Avoid unnecessary empty lines

File: catkin_ws/src/image_rotation/include/image_rotation/ImageAlgorithm.h (hunk 1 above)
Comment (AshAswin) [inline catkin_ws/src/image_rotation/include/image_rotation/ImageAlgorithm.h]: Provide a file description

File: catkin_ws/src/image_rotation/src/ImageAlgorithm.cpp [hunk 4]
```diff
@@ -0,0 +1,16 @@
+#include "ImageAlgorithm.h"
+
+void ImageAlgorithm::RotateImage(cv::Mat &image, double angle)
+{
+    // get rotation matrix for rotating the image around its center in pixel coordinates
+    cv::Point2f center((image.cols - 1) / 2.0, (image.rows - 1) / 2.0);
+    cv::Mat rot = cv::getRotationMatrix2D(center, angle, 1.0);
```
Comment (AshAswin) [inline catkin_ws/src/image_rotation/src/ImageAlgorithm.cpp]: Use meaningful object names. rot -> rotation_matrix

File: catkin_ws/src/image_rotation/include/image_rotation/ImageAlgorithm.h [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+{
+  public:
+    ImageAlgorithm()
+    {
+    }
+    static void RotateImage(cv::Mat &image, double angle);
+
+};
+
+#endif
```
Comment (AshAswin) [inline catkin_ws/src/image_rotation/include/image_rotation/ImageAlgorithm.h]: EOL character missing. Comment for header guards missing

File: catkin_ws/src/image_rotation/src/ImageAlgorithm.cpp [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+    // determine bounding rectangle, center not relevant
+    cv::Rect2f bbox = cv::RotatedRect(cv::Point2f(), image.size(), angle).boundingRect2f();
```
Comment (AshAswin) [inline catkin_ws/src/image_rotation/src/ImageAlgorithm.cpp]: meaningful object name for bbox

File: catkin_ws/src/image_rotation/include/image_rotation/ImageAlgorithm.h [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+#define IMAGE_ALGORITHM_H
+
+#include <opencv2/opencv.hpp>
+
+
+class ImageAlgorithm
+{
+  public:
+    ImageAlgorithm()
+    {
```
Comment (AshAswin) [inline catkin_ws/src/image_rotation/include/image_rotation/ImageAlgorithm.h]: Avoid empty constructors. Use  default keyword

**Review by AshAswin — CHANGES_REQUESTED** [review]
Please make the requested changes

File: catkin_ws/src/image_rotation/include/image_rotation/ImageAlgorithm.h (hunk 1 above)
Comment (upengareri) [reply catkin_ws/src/image_rotation/include/image_rotation/ImageAlgorithm.h]: fixed.

File: catkin_ws/src/image_rotation/include/image_rotation/ImageAlgorithm.h (hunk 2 above)
Comment (upengareri) [reply catkin_ws/src/image_rotation/include/image_rotation/ImageAlgorithm.h]: we can look into this later.

File: catkin_ws/src/image_rotation/include/image_rotation/ImageAlgorithm.h (hunk 3 above)
Comment (upengareri) [reply catkin_ws/src/image_rotation/include/image_rotation/ImageAlgorithm.h]: fixed

File: catkin_ws/src/image_rotation/include/image_rotation/ImageAlgorithm.h (hunk 1 above)
Comment (upengareri) [reply catkin_ws/src/image_rotation/include/image_rotation/ImageAlgorithm.h]: done

File: catkin_ws/src/image_rotation/include/image_rotation/ImageAlgorithm.h (hunk 3 above)
Comment (upengareri) [reply catkin_ws/src/image_rotation/include/image_rotation/ImageAlgorithm.h]: fixed

File: catkin_ws/src/image_rotation/src/ImageAlgorithm.cpp (hunk 4 above)
Comment (upengareri) [reply catkin_ws/src/image_rotation/src/ImageAlgorithm.cpp]: fixed

File: catkin_ws/src/image_rotation/include/image_rotation/ImageAlgorithm.h (hunk 5 above)
Comment (upengareri) [reply catkin_ws/src/image_rotation/include/image_rotation/ImageAlgorithm.h]: fixed

File: catkin_ws/src/image_rotation/src/ImageAlgorithm.cpp (hunk 6 above)
Comment (upengareri) [reply catkin_ws/src/image_rotation/src/ImageAlgorithm.cpp]: I feel bbox is universal naming trend and if you search the keyword in reference to image processing you will get the desired result

File: catkin_ws/src/image_rotation/include/image_rotation/ImageAlgorithm.h (hunk 7 above)
Comment (upengareri) [reply catkin_ws/src/image_rotation/include/image_rotation/ImageAlgorithm.h]: fixed

File: catkin_ws/src/image_rotation/src/ImageAlgorithm.cpp (hunk 6 above)
Comment (aravindk2604) [reply catkin_ws/src/image_rotation/src/ImageAlgorithm.cpp]: in BMW adp code base they use _bounding_box_ instead of bbox

**Comment by aravindk2604** [comment]
Thanks @AshAswin  for the keen review. I have no more additions from my side.

File: catkin_ws/src/image_rotation/src/image_algorithm.cpp [hunk 8]
```diff
@@ -0,0 +1,23 @@
+///
+/// @file
+/// @brief Implementation of image processing class
```
Comment (AshAswin) [inline catkin_ws/src/image_rotation/src/image_algorithm.cpp]: mistake in name of the class

File: catkin_ws/src/image_rotation/src/image_algorithm.cpp [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+///
+
+#include "ImageAlgorithm.h"
```
Comment (AshAswin) [inline catkin_ws/src/image_rotation/src/image_algorithm.cpp]: mistake in name of the header file. The code doesn't compile

File: catkin_ws/src/image_rotation/src/image_algorithm.cpp [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+
+using namespace cv;
+
+void ImageAlgorithm::RotateImage(Mat &image, double angle)
+{
+    // get rotation matrix for rotating the image around its center in pixel coordinates
+    Point2f center((image.cols - 1) / 2.0, (image.rows - 1) / 2.0);
+    Mat rotation_matrix = getRotationMatrix2D(center, angle, 1.0);
+    // determine bounding rectangle, center not relevant
+    Rect2f bbox = RotatedRect(Point2f(), image.size(), angle).boundingRect2f();
```
Comment (AshAswin) [inline catkin_ws/src/image_rotation/src/image_algorithm.cpp]: change variable name from bbox to bounding_box

**Review by AshAswin — CHANGES_REQUESTED** [review]
Changes required

File: catkin_ws/src/image_rotation/src/image_algorithm.cpp (hunk 8 above)
Comment (upengareri) [reply catkin_ws/src/image_rotation/src/image_algorithm.cpp]: fixed

File: catkin_ws/src/image_rotation/src/image_algorithm.cpp (hunk 9 above)
Comment (upengareri) [reply catkin_ws/src/image_rotation/src/image_algorithm.cpp]: my bad - has been fixed

File: catkin_ws/src/image_rotation/src/image_algorithm.cpp (hunk 10 above)
Comment (upengareri) [reply catkin_ws/src/image_rotation/src/image_algorithm.cpp]: changed

File: catkin_ws/src/image_rotation/src/image_algorithm.cpp [hunk 11]
```diff
@@ -0,0 +1,23 @@
+///
+/// @file
+/// @brief Implementation of image processing algorithm
+///
+
+#include "image_algorithm.h"
+
+using namespace cv;
```
Comment (aravindk2604) [inline catkin_ws/src/image_rotation/src/image_algorithm.cpp:8]: @melisamariam 
According to [Namespaces in Google Style guide C++](https://google.github.io/styleguide/cppguide.html#Namespaces), it is forbidden to use "using namespace foo;"
@upengareri : please make the necessary changes

**Review by AshAswin — APPROVED** [review]

**Pull request merged**